src/NUnitFramework/framework/Internal/Execution/WorkShift.cs

```diff
@@ -161,16 +170,18 @@ private void StartWorkers()
                 worker.Idle += (s, ea) =>
                 {
                     // Quick check first using Interlocked.Decrement
-                    if (Interlocked.Decrement(ref _busyCount) == 0)
+                    if (Interlocked.Decrement(ref _busyCount) == 0 && !HasWork)
+                    {
                         lock (_syncRoot)
                         {
-                            // Check busy count again under the lock. If there is no work
-                            // try to restore any saved queues and end the shift.
+                            // Check again under the lock. If there is no work
```

Comment:
@jnm2 said above,

> having more than one concurrency primitive in action makes me feel like everything is susceptible to race conditions. What happens if one of the queues gets an item added on a background thread right after IsEmpty returns false?

@jnm2 isn't this okay because we check `_busyCount` and `HasWork` again after getting the lock? The comment is pretty obvious, so what you are seeing is more subtle than that. Can you expand on your concern?

Reply:
I agree with both of you. I'm worried like @jnm2 and not seeing how it happens like @rprouse. 😄 

My observation is that things get in - it's rare but happens.